### PR TITLE
Detecting redeclaration of p5.js reserved constants and functions

### LINKED ISSUE #5350: GSoC'21(FES):  Detecting reassignment of p5.js reserved constants and functions
### How would this new feature help [increase access](https://github.com/processing/p5.js/blob/main/contributor_docs/access.md) to p5.js?
FES is not able to detect if the user has reassigned p5.js reserved constants or functions this can create some problems as the default values of the constants/functions would be overwritten. This feature aims to prevent that from happening.

#### Most appropriate sub-area of p5.js?

- [ ] Accessibility (Web Accessibility)
- [ ] Build tools and processes
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [x] Friendly error system
- [ ] Image
- [ ] IO (Input/Output)
- [ ] Localization
- [ ] Math
- [ ] Unit Testing
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Other (specify if possible)

### PR DESCRIPTION
**Pull request by Aloneduckling**
Resolves #5350

 Changes:

- The new file `sketch_reader.js` contains the code that will help the FES detect redeclaration of p5.js constants and functions.
- Inline documentation is included
- Unit tests are included

 Screenshots of the change:

Detection of p5.js reserved function:
image

Detection of p5.js reserved constant
image

#### PR Checklist

- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated
- [x] [Unit tests] are included / updated

[Inline documentation]: https://github.com/processing/p5.js/blob/main/contributor_docs/inline_documentation.md
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

### CODE REVIEW

**Comment by Aloneduckling** [comment]
This is a draft PR

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:175]: throwing an error here: "Uncaught (in promise) TypeError: Failed to fetch"

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:181]: same fetching error here

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:144]: I'm getting a fetching error in this promise

File: translations/en/translation.json
Comment (lm-n) [reply translations/en/translation.json:60]: I wonder how these compares to the style of other errors.  let chat about it!

**Review by lm-n — CHANGES_REQUESTED** [review]
Great work! It is throwing some fetching errors when running locally and a "Range Error" on the editor. I have a couple of questions about style of the messages that we can talk about. It might be good to start adding tests to catch errors sooner than later.

File: translations/en/translation.json
Comment (Aloneduckling) [reply translations/en/translation.json:60]: yeah cool!

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [inline src/core/friendly_errors/sketch_reader.js:271]: Nice, great job!

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [inline src/core/friendly_errors/sketch_reader.js:148]: Looks good! Does `ele.match(reg)` always have a `[1]` element?

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [inline src/core/friendly_errors/sketch_reader.js:124]: I suggest splitting this code a bit: You can try creating a function from the lambda at line 104, and end up with something like:

```javascript
const extractVariables = arr => arr
        .map(applyVariableRegex) // Apply regexes to check for matches (lines 104 to 113, kind of)
        .filter(...) // Filter out whatever doesn't match (line 113)
        .reduce(...) // Flatten the "ele" items into the final list

const applyVariableRegex = s => {
  if (s.includes('=')) { ... }
  ...
}

...
```

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [inline src/core/friendly_errors/sketch_reader.js:245]: You can use `Array.every()` here:
```javascript

ignore_files = ['p5.js', 'p5.min.js', 'p5.sounds.js', 'p5.sounds.min.js', 'previewScripts']

...
  .filter(attr => attr !== null
      ...
      && ignore_files.every(filename => !attr.includes(filename))
  )

```

**Review by ghalestrilo — CHANGES_REQUESTED** [review]
I think this is coming out great. Some parts of the code can be improved but I think the most important is the analysis logic: there are some cases the code doesn't catch yet

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:181]: yeah, I will update the code.

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:148]: yes, it is because we are using capture groups here and they return an array with 2 elements in them and the 2nd element is the one that we require.

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:124]: the `extractVariables` function's sole purpose is to apply the regex and extract the function names so I guess code should be here and using another function for regex matching would be overdoing it. I am happy to defer though. What are your thoughts?

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:171]: I have updated the code

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [reply src/core/friendly_errors/sketch_reader.js]: do this for all variables across the file: for example instead of ele say element or whatever ele stands for. instead of arr say elementArray

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:99]: great in line documentation here! Can you do the same from lines 68-78

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:94]: `keyArray` instead of `keyArr`

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:133]: for all regex add a comment describing what you are doing

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:126]: use descriptive variable names

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [inline src/core/friendly_errors/sketch_reader.js:222]: use descriptive variable name instead of str

**Review by lm-n — CHANGES_REQUESTED** [review]
Great progress! My main comment is please use descriptive variable names even if they are verbose. We do this to ensure human readability of the code particularly for novice contributors. We cannot have arrays called `arr` or strings called `str`. Also for variables like elements give a descriptive name.

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [reply src/core/friendly_errors/sketch_reader.js]: I agree with @Aloneduckling on this one!

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js]: ok I will update the code

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:68]: ok I will make the changes soon

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [reply src/core/friendly_errors/sketch_reader.js]: Alright! That makes sense

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [inline src/core/friendly_errors/sketch_reader.js:349]: I think this line is a bit confusing - using `isFound` to break the loop is clearer (if that's possible). In fact, you could just `return true` straight from line 344

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [inline src/core/friendly_errors/sketch_reader.js:114]: Same as line `349`

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [inline src/core/friendly_errors/sketch_reader.js:324]: I don't see the index `i` being used anywhere else, maybe you can use a for-of instead:
```javascript
for (let function_name of functionArray) { ... }
```

**Review by ghalestrilo — CHANGES_REQUESTED** [review]
I made some minor code suggestions, see what you think

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:324]: I used for loop here because when p5.js is bundled it is converted to ES5 format and this conversion sometimes causes problems especially with the new variations of for loop. So to be on the safer side I used the traditional for loop.

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:349]: Great suggestion I will make the changes

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:372]: I will make the changes

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:349]: This is a great suggestion thanks, I will change the code and push it soon

File: src/core/friendly_errors/sketch_reader.js
Comment (Aloneduckling) [reply src/core/friendly_errors/sketch_reader.js:152]: Thanks for the suggestion, will make the changes.

File: src/core/friendly_errors/sketch_reader.js
Comment (ghalestrilo) [reply src/core/friendly_errors/sketch_reader.js:324]: I see, that makes sense. I'll mark this as resolved then

File: src/core/friendly_errors/sketch_reader.js
Comment (lm-n) [reply src/core/friendly_errors/sketch_reader.js:145]: awesome!

**Review by lm-n — COMMENTED** [review]
Great progress! Good job documenting the sketch_reader.js. Let's wait for @ghalestrilo feedback and then we can see if we're ready to merge :)

File: contributor_docs/friendly_error_system.md
Comment (lm-n) [inline contributor_docs/friendly_error_system.md:248]: be more detailed about what you mean with this

File: contributor_docs/friendly_error_system.md
Comment (lm-n) [inline contributor_docs/friendly_error_system.md:249]: new features such as

**Review by lm-n — COMMENTED** [review]
Please make some edits to limitations and thoughts for the future. Try to be as specific as you can.

File: contributor_docs/friendly_error_system.md
Comment (Aloneduckling) [reply contributor_docs/friendly_error_system.md]: I made the changes. Please have a look

**Review by ghalestrilo — APPROVED** [review]
Nice! I don't see anything that has to change now before merging.
With the updates to the documentation I think this branch is ready
Great job on the tests!

**Review by lm-n — APPROVED** [review]

**Pull request merged**
